src/config/enterprise.rs

```diff
@@ -0,0 +1,426 @@
+                                    break;
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
+        // If we're at this point, there was an error that was deemed non-fatal, so retry.
+        info!(
+            "Retrying config reporting to {} in {} seconds.",
+            DATADOG_REPORTING_PRODUCT, datadog.retry_interval_secs
+        );
+
+        // Sleep for the user-provided interval before retrying.
+        sleep(Duration::from_secs(datadog.retry_interval_secs as u64)).await;
+    }
+
+    let host_metrics_id = OutputId::from(ComponentKey::from(HOST_METRICS_KEY));
+    let internal_metrics_id = OutputId::from(ComponentKey::from(INTERNAL_METRICS_KEY));
+    let datadog_metrics_id = ComponentKey::from(DATADOG_METRICS_KEY);
+
+    // Create internal sources for host and internal metrics. We're using distinct sources here and
+    // not attempting to reuse existing ones, to configure according to enterprise requirements.
+    let mut host_metrics =
+        HostMetricsConfig::enterprise(config_version, &datadog.configuration_key);
+    let mut internal_metrics =
+        InternalMetricsConfig::enterprise(config_version, &datadog.configuration_key);
+
+    // Override default scrape intervals.
+    host_metrics.scrape_interval_secs(datadog.reporting_interval_secs);
+    internal_metrics.scrape_interval_secs(datadog.reporting_interval_secs);
+
+    config.sources.insert(
+        host_metrics_id.component.clone(),
+        SourceOuter::new(host_metrics),
+    );
+    config.sources.insert(
+        internal_metrics_id.component.clone(),
+        SourceOuter::new(internal_metrics),
+    );
+
+    // Create a Datadog metrics sink to consume and emit internal + host metrics.
+    let datadog_metrics = DatadogMetricsConfig::from_api_key(api_key);
+
+    config.sinks.insert(
+        datadog_metrics_id,
+        SinkOuter::new(
+            vec![host_metrics_id, internal_metrics_id],
+            Box::new(datadog_metrics),
+        ),
+    );
+
+    Ok(())
+}
+
+/// By default, the Datadog feature is enabled.
+const fn default_enabled() -> bool {
+    true
+}
+
+/// By default, Vector should exit when a fatal reporting error is encountered.
+const fn default_exit_on_fatal_error() -> bool {
+    true
+}
+
+/// By default, report to Datadog every 5 seconds.
+const fn default_reporting_interval_secs() -> f64 {
+    5.0
+}
+
+/// By default, keep retrying (recoverable) failed reporting (infinitely, for practical purposes.)
+const fn default_max_retries() -> u32 {
+    u32::MAX
+}
+
+/// By default, retry (recoverable) failed reporting every 5 seconds.
+const fn default_retry_interval_secs() -> u32 {
+    5
+}
+
+/// Returns the full URL endpoint of where to POST a Datadog Vector configuration.
+fn get_reporting_endpoint(site: Option<&String>, configuration_key: &str) -> String {
```

Comment:
Could this return a `Url` instead of a `String`? (to avoid the potentially repeated `Url::parse` in `report_serialized_config_to_datadog`)